engine/src/main/java/org/hibernate/validator/internal/metadata/aggregated/CascadingMetaDataBuilder.java

```diff
@@ -317,40 +321,37 @@ public boolean equals(Object obj) {
 		return mergedGroupConversions;
 	}
 
-	private static Map<TypeVariable<?>, CascadingMetaDataBuilder> addPotentialLegacyCascadingMetaData(Type cascadableType,
-			Map<TypeVariable<?>, CascadingMetaDataBuilder> containerElementTypesCascadingMetaData, Map<Class<?>, Class<?>> groupConversions) {
+	private static Map<TypeVariable<?>, CascadingMetaDataBuilder> addPotentialCascadingMetaData(Type cascadableType, Map<TypeVariable<?>, CascadingMetaDataBuilder> containerElementTypesCascadingMetaData, Map<Class<?>, Class<?>> groupConversions,
+			Map<TypeVariable<?>, ValueExtractorDescriptor> possibleValueExtractorCandidatesForCascadedValidation) {
 		Class<?> cascadableClass = ReflectionHelper.getClassFromType( cascadableType );
-
-		if ( Map.class.isAssignableFrom( cascadableClass ) ) {
-			return addLegacyCascadingMetaData( cascadableClass, Map.class, 1, containerElementTypesCascadingMetaData, groupConversions );
-		}
-		else if ( List.class.isAssignableFrom( cascadableClass ) ) {
-			return addLegacyCascadingMetaData( cascadableClass, List.class, 0, containerElementTypesCascadingMetaData, groupConversions );
-		}
-		else if ( Iterable.class.isAssignableFrom( cascadableClass ) ) {
-			return addLegacyCascadingMetaData( cascadableClass, Iterable.class, 0, containerElementTypesCascadingMetaData, groupConversions );
-		}
-		else if ( Optional.class.isAssignableFrom( cascadableClass ) ) {
-			return addLegacyCascadingMetaData( cascadableClass, Optional.class, 0, containerElementTypesCascadingMetaData, groupConversions );
-		}
-		else if ( cascadableClass.isArray() ) {
+		if ( cascadableClass.isArray() ) {
 			// for arrays, we need to add an ArrayElement cascading metadata: it's the only way arrays support cascading at the moment.
 			return addArrayElementCascadingMetaData( cascadableClass, containerElementTypesCascadingMetaData, groupConversions );
 		}
 		else {
-			return containerElementTypesCascadingMetaData;
+			Map<TypeVariable<?>, CascadingMetaDataBuilder> cascadingMetaData = containerElementTypesCascadingMetaData;
+			for ( Entry<TypeVariable<?>, ValueExtractorDescriptor> extractorDescriptorEntry : possibleValueExtractorCandidatesForCascadedValidation.entrySet() ) {
```

Comment:
Does it really work? I'm especially surprised that we might have several VED on different type arguments. I don't see how this could work.

If we have at least 2 VED for different type arguments, we will throw an error at one point, don't we?

Reply:
I was referring to this comment.